File: sdk/java/src/main/java/io/agentsafe/spl/Crypto.java
package io.agentsafe.spl;

import java.security.*;
import java.security.spec.EdECPublicKeySpec;
import java.security.spec.NamedParameterSpec;
import java.util.Arrays;

/**
 * Crypto implementations for SPL token verification.
 * Ed25519 via java.security.Signature (EdDSA, JDK 17+).
 * SHA-256 via MessageDigest.
 */
public final class Crypto {
    private Crypto() {}

    /**
     * Verify an Ed25519 signature over a message.
     */
    public static boolean verifyEd25519(byte[] message, String signatureHex, String publicKeyHex) {
        try {
            byte[] sigBytes = hexToBytes(signatureHex);
            byte[] pubBytes = hexToBytes(publicKeyHex);
            if (sigBytes.length != 64 || pubBytes.length != 32) return false;

            // Construct the public key using EdDSA
            KeyFactory kf = KeyFactory.getInstance("EdDSA");
            // Ed25519 public key from raw bytes
            // Use X509EncodedKeySpec with DER prefix
            byte[] derPrefix = hexToBytes("302a300506032b6570032100");
            byte[] encoded = new byte[derPrefix.length + pubBytes.length];
            System.arraycopy(derPrefix, 0, encoded, 0, derPrefix.length);
            System.arraycopy(pubBytes, 0, encoded, derPrefix.length, pubBytes.length);

            PublicKey publicKey = kf.generatePublic(new java.security.spec.X509EncodedKeySpec(encoded));
            Signature sig = Signature.getInstance("EdDSA");
            sig.initVerify(publicKey);
            sig.update(message);
            return sig.verify(sigBytes);
        } catch (Exception e) {
            return false;
        }
    }

    /**
     * SHA-256 hash of data.
     */
    public static byte[] sha256(byte[] data) {
        try {
            MessageDigest md = MessageDigest.getInstance("SHA-256");
            return md.digest(data);
        } catch (NoSuchAlgorithmException e) {
            throw new RuntimeException("SHA-256 not available", e);
        }
    }

    /**
     * SHA-256 hash of data as hex string.
     */
    public static String sha256Hex(byte[] data) {
        return bytesToHex(sha256(data));
    }

    /**
     * Verify a Merkle proof.
     */
    public static boolean verifyMerkleProof(String leafData, String[][] proof, String rootHex) {
        try {
            byte[] current = sha256(leafData.getBytes(java.nio.charset.StandardCharsets.UTF_8));
            for (String[] step : proof) {
                String hash = step[0];
                String position = step[1];
                byte[] sibling = hexToBytes(hash);
                MessageDigest md = MessageDigest.getInstance("SHA-256");
                if (position.equals("right")) {
                    md.update(current);
                    md.update(sibling);
                } else {
                    md.update(sibling);
                    md.update(current);
                }
                current = md.digest();
            }
            return bytesToHex(current).equals(rootHex);
        } catch (Exception e) {
            return false;
        }
    }

    /**
     * Verify a hash chain receipt.
     */
    public static boolean verifyHashChain(String commitment, String preimageHex, int index, int chainLength) {
        try {
            byte[] current = hexToBytes(preimageHex);
            int steps = chainLength - index;
            for (int i = 0; i < steps; i++) {
                current = sha256(current);
            }
            return bytesToHex(current).equals(commitment);
        } catch (Exception e) {
            return false;
        }
    }

    /**
     * HMAC-SHA256.
     */
    private static byte[] hmacSha256(byte[] key, byte[] data) {
        try {
            javax.crypto.Mac mac = javax.crypto.Mac.getInstance("HmacSHA256");
            mac.init(new javax.crypto.spec.SecretKeySpec(key, "HmacSHA256"));
            return mac.doFinal(data);
        } catch (Exception e) {
            throw new RuntimeException("HMAC-SHA256 not available", e);
        }
    }

    /**
     * HKDF-SHA256 (RFC 5869) extract-and-expand. Zero external dependencies.
     */
    private static byte[] hkdfSha256(byte[] ikm, byte[] salt, byte[] info, int length) {
        // Extract
        if (salt == null || salt.length == 0) salt = new byte[32];
        byte[] prk = hmacSha256(salt, ikm);

        // Expand
        byte[] out = new byte[0];
        byte[] prev = new byte[0];
        for (int i = 1; out.length < length; i++) {
            byte[] input = new byte[prev.length + info.length + 1];
            System.arraycopy(prev, 0, input, 0, prev.length);
            System.arraycopy(info, 0, input, prev.length, info.length);
            input[input.length - 1] = (byte) i;
            prev = hmacSha256(prk, input);
            byte[] newOut = new byte[out.length + prev.length];
            System.arraycopy(out, 0, newOut, 0, out.length);
            System.arraycopy(prev, 0, newOut, out.length, prev.length);
            out = newOut;
        }
        return java.util.Arrays.copyOf(out, length);
    }

    /**
     * Derive a service-specific Ed25519 keypair using HKDF-SHA256.
     * Provides unlinkability: different services see different public keys.
     *
     * @return [publicKeyHex, privateKeyHex]
     */
    public static String[] deriveServiceKey(String masterKeyHex, String serviceDomain) {
        try {
            byte[] masterKey = hexToBytes(masterKeyHex);
            byte[] salt = "agent-safe-v1".getBytes(java.nio.charset.StandardCharsets.UTF_8);
            byte[] info = serviceDomain.getBytes(java.nio.charset.StandardCharsets.UTF_8);
            byte[] seed = hkdfSha256(masterKey, salt, info, 32);

            // Build PKCS8 DER for Ed25519 private key from seed
            byte[] pkcs8Prefix = hexToBytes("302e020100300506032b657004220420");
            byte[] privDer = new byte[pkcs8Prefix.length + seed.length];
            System.arraycopy(pkcs8Prefix, 0, privDer, 0, pkcs8Prefix.length);
            System.arraycopy(seed, 0, privDer, pkcs8Prefix.length, seed.length);

            KeyFactory kf = KeyFactory.getInstance("Ed25519");
            java.security.PrivateKey privKey = kf.generatePrivate(
                new java.security.spec.PKCS8EncodedKeySpec(privDer));

            // Extract public key
            java.security.interfaces.EdECPrivateKey edPriv =
                (java.security.interfaces.EdECPrivateKey) privKey;
            byte[] encoded = edPriv.getEncoded();
            // Look for public key in PKCS8 v2 encoding
            for (int i = 0; i < encoded.length - 33; i++) {
                if (encoded[i] == (byte) 0xa1 && encoded[i + 1] == (byte) 0x23
                    && encoded[i + 2] == (byte) 0x03 && encoded[i + 3] == (byte) 0x21
                    && encoded[i + 4] == (byte) 0x00) {
                    byte[] pubBytes = new byte[32];
                    System.arraycopy(encoded, i + 5, pubBytes, 0, 32);

                    boolean xOdd = (pubBytes[31] & 0x80) != 0;
                    pubBytes[31] &= 0x7f;
                    byte[] reversed = new byte[32];
                    for (int j = 0; j < 32; j++) reversed[j] = pubBytes[31 - j];
                    java.math.BigInteger y = new java.math.BigInteger(1, reversed);
                    java.security.spec.EdECPoint pubPoint =
                        new java.security.spec.EdECPoint(xOdd, y);
                    java.security.spec.EdECPublicKeySpec pubSpec =
                        new java.security.spec.EdECPublicKeySpec(
                            java.security.spec.NamedParameterSpec.ED25519, pubPoint);
                    java.security.PublicKey pubKey = kf.generatePublic(pubSpec);
                    byte[] pubDer = pubKey.getEncoded();
                    byte[] pubRaw = new byte[32];
                    System.arraycopy(pubDer, pubDer.length - 32, pubRaw, 0, 32);
                    return new String[]{bytesToHex(pubRaw), bytesToHex(seed)};
                }
            }
            throw new SplException("Could not extract public key from derived private key");
        } catch (SplException e) {
            throw e;
        } catch (Exception e) {
            throw new SplException("Failed to derive service key: " + e.getMessage());
        }
    }

    public static byte[] hexToBytes(String hex) {
        int len = hex.length();
        byte[] data = new byte[len / 2];
        for (int i = 0; i < len; i += 2) {
            data[i / 2] = (byte) ((Character.digit(hex.charAt(i), 16) << 4)
                + Character.digit(hex.charAt(i + 1), 16));
        }
        return data;
    }

    public static String bytesToHex(byte[] bytes) {
        StringBuilder sb = new StringBuilder(bytes.length * 2);
        for (byte b : bytes) {
            sb.append(String.format("%02x", b & 0xff));
        }
        return sb.toString();
    }
}

File: sdk/csharp/AgentSafe.Spl/Crypto.cs
using System.Security.Cryptography;
// HMACSHA256 is in System.Security.Cryptography

namespace AgentSafe.Spl;

/// <summary>
/// Crypto implementations for SPL token verification.
/// SHA-256 Merkle + hash-chain via stdlib.
/// Ed25519 available in .NET 9+ via Ed25519.
/// </summary>
public static class Crypto
{
    public static byte[] Sha256(byte[] data) => SHA256.HashData(data);

    public static string Sha256Hex(byte[] data) => Convert.ToHexString(Sha256(data)).ToLowerInvariant();

    private static string ToHexLower(byte[] data) => Convert.ToHexString(data).ToLowerInvariant();

    /// <summary>Verify a Merkle proof for leafData against rootHex.</summary>
    public static bool VerifyMerkleProof(string leafData, (string Hash, string Position)[] proof, string rootHex)
    {
        try
        {
            var current = Sha256(System.Text.Encoding.UTF8.GetBytes(leafData));
            foreach (var (hash, position) in proof)
            {
                var sibling = Convert.FromHexString(hash);
                using var sha = IncrementalHash.CreateHash(HashAlgorithmName.SHA256);
                if (position == "right")
                {
                    sha.AppendData(current);
                    sha.AppendData(sibling);
                }
                else
                {
                    sha.AppendData(sibling);
                    sha.AppendData(current);
                }
                current = sha.GetHashAndReset();
            }
            return ToHexLower(current) == rootHex;
        }
        catch
        {
            return false;
        }
    }

    /// <summary>Verify a hash chain receipt.</summary>
    public static bool VerifyHashChain(string commitment, string preimageHex, int index, int chainLength)
    {
        try
        {
            var current = Convert.FromHexString(preimageHex);
            var steps = chainLength - index;
            for (var i = 0; i < steps; i++)
                current = Sha256(current);
            return ToHexLower(current) == commitment;
        }
        catch
        {
            return false;
        }
    }

    /// <summary>HKDF-SHA256 (RFC 5869) extract-and-expand. Zero external dependencies.</summary>
    private static byte[] HkdfSha256(byte[] ikm, byte[] salt, byte[] info, int length)
    {
        if (salt == null || salt.Length == 0) salt = new byte[32];
        // Extract: PRK = HMAC-SHA256(salt, ikm)
        using var extractMac = new HMACSHA256(salt);
        var prk = extractMac.ComputeHash(ikm);

        // Expand
        var result = new byte[length];
        var prev = Array.Empty<byte>();
        var offset = 0;
        for (byte i = 1; offset < length; i++)
        {
            using var expandMac = new HMACSHA256(prk);
            expandMac.TransformBlock(prev, 0, prev.Length, null, 0);
            expandMac.TransformBlock(info, 0, info.Length, null, 0);
            expandMac.TransformFinalBlock(new[] { i }, 0, 1);
            prev = expandMac.Hash!;
            var toCopy = Math.Min(prev.Length, length - offset);
            Array.Copy(prev, 0, result, offset, toCopy);
            offset += toCopy;
        }
        return result;
    }

    /// <summary>
    /// Derive a service-specific Ed25519 seed using HKDF-SHA256.
    /// Returns (publicKeyHex, privateKeyHex) when Ed25519 is available.
    /// Without AGENTSAFE_ED25519, returns only the derived seed hex.
    /// </summary>
    public static (string Seed, string? PublicKey) DeriveServiceSeed(string masterKeyHex, string serviceDomain)
    {
        var masterKey = Convert.FromHexString(masterKeyHex);
        var salt = System.Text.Encoding.UTF8.GetBytes("agent-safe-v1");
        var info = System.Text.Encoding.UTF8.GetBytes(serviceDomain);
        var seed = HkdfSha256(masterKey, salt, info, 32);
        var seedHex = Convert.ToHexString(seed).ToLowerInvariant();
#if AGENTSAFE_ED25519
        // When Ed25519 is available, derive the full keypair
        // For now, return just the seed
#endif
        return (seedHex, null);
    }

// Ed25519 is not yet available in the .NET standard library.
// When it becomes available, uncomment the following:
#if AGENTSAFE_ED25519
    /// <summary>Verify an Ed25519 signature (requires .NET 9+).</summary>
    public static bool VerifyEd25519(byte[] message, string signatureHex, string publicKeyHex)
    {
        try
        {
            var sig = Convert.FromHexString(signatureHex);
            var pub = Convert.FromHexString(publicKeyHex);
            if (sig.Length != 64 || pub.Length != 32) return false;
            return System.Security.Cryptography.Ed25519.Verify(pub, message, sig);
        }
        catch
        {
            return false;
        }
    }
#endif
}
